src/modules/roc_netio/target_uv/roc_netio/udp_receiver.cpp

```diff
@@ -42,59 +44,20 @@ void UDPReceiver::destroy() {
 }
 
 bool UDPReceiver::start(packet::Address& bind_address) {
-    if (int err = uv_udp_init(&loop_, &handle_)) {
-        roc_log(LogError, "udp receiver: uv_udp_init(): [%s] %s", uv_err_name(err),
-                uv_strerror(err));
-        return false;
-    }
-
-    handle_.data = this;
-    handle_initialized_ = true;
-
-    unsigned flags = 0;
-    if (bind_address.multicast() && bind_address.port() > 0) {
-        flags |= UV_UDP_REUSEADDR;
-    }
-
-    int bind_err = UV_EINVAL;
-    if (bind_address.version() == 6) {
-        bind_err = uv_udp_bind(&handle_, bind_address.saddr(), flags | UV_UDP_IPV6ONLY);
-    }
-    if (bind_err == UV_EINVAL || bind_err == UV_ENOTSUP) {
-        bind_err = uv_udp_bind(&handle_, bind_address.saddr(), flags);
-    }
-    if (bind_err != 0) {
-        roc_log(LogError, "udp receiver: uv_udp_bind(): [%s] %s", uv_err_name(bind_err),
-                uv_strerror(bind_err));
-        return false;
-    }
-
-    int addrlen = (int)bind_address.slen();
-    if (int err = uv_udp_getsockname(&handle_, bind_address.saddr(), &addrlen)) {
-        roc_log(LogError, "udp receiver: uv_udp_getsockname(): [%s] %s", uv_err_name(err),
-                uv_strerror(err));
-        return false;
-    }
+    const bool started = start_(bind_address);
+    address_ = bind_address;
 
-    if (addrlen != (int)bind_address.slen()) {
-        roc_log(
-            LogError,
-            "udp receiver: uv_udp_getsockname(): unexpected len: got=%lu expected=%lu",
-            (unsigned long)addrlen, (unsigned long)bind_address.slen());
-        return false;
+    if (started) {
+        return true;
     }
 
-    if (int err = uv_udp_recv_start(&handle_, alloc_cb_, recv_cb_)) {
-        roc_log(LogError, "udp receiver: uv_udp_recv_start(): [%s] %s", uv_err_name(err),
-                uv_strerror(err));
-        return false;
+    if (handle_initialized_) {
+        stop();
+    } else {
+        close_handler_.close(*this);
     }
```

Comment:
It probably makes sense to move this code:

```
    if (handle_initialized_) {
        stop();
    } else {
        close_handler_.close(*this);
    }
```

from UDPSender and UDPReceiver to Transceiver by requiring it to call stop() even when start() failed:

```
if (!rp->start()) {
    rp->stop();
    return false;
}
```

and by adding this to stop() implementation:

```
if (!handle_initialized_) {
    close_handler_.close(*this);
}
```

Why? It's quite expected that ICloseHandler may be called iniside UDPSender::stop() (at least if we'll rename it to async_close()) but it's quite unexpected that it may be called inside UDPSender::start().

Reply:
Probably, it was done in a such non-obvious way because it is a little bit
strange to call stop() on non-started object.

Reply:
Also, it was complicated because it was required to initialize UDPSender(Receiver)
address even if the start() was failed (due to stop logic that depends on the address).
Now we put an object directly to callback and do not have such problems.

Reply:
Reworked